File: fsharp/space-age/SpaceAge.fs
module SpaceAge
(*
https://stackoverflow.com/questions/12156022/defining-modules-vs-net-vs-f-interactive
#if INTERACTIVE
#else
module SpaceAge
#endif
*)

type Planet =
    | Venus
    | Mercury
    | Earth
    | Mars
    | Jupiter
    | Saturn
    | Uranus
    | Neptune

let age (planet: Planet) (seconds: int64) : float =
    let earthYear = 31557600L |> float

    let planetYear =
        match planet with
        | Mercury -> 0.2408467
        | Venus -> 0.61519726
        | Earth -> 1.
        | Mars -> 1.8808158
        | Jupiter -> 11.862615
        | Saturn -> 29.447498
        | Uranus -> 84.016846
        | Neptune -> 164.79132

    (seconds |> float) / (planetYear * earthYear)

File: elixir/space-age/lib/space_age.ex
defmodule SpaceAge do
  @type planet ::
          :mercury
          | :venus
          | :earth
          | :mars
          | :jupiter
          | :saturn
          | :uranus
          | :neptune

  @doc """
  Return the number of years a person that has lived for 'seconds' seconds is
  aged on 'planet', or an error if 'planet' is not a planet.
  """
  @spec age_on(planet, pos_integer) :: {:ok, float} | {:error, String.t()}
  def age_on(:mercury, seconds), do: {:ok, earth_age(seconds) / 0.2408467}
  def age_on(:venus, seconds), do: {:ok, earth_age(seconds) / 0.61519726}
  def age_on(:earth, seconds), do: {:ok, earth_age(seconds)}
  def age_on(:mars, seconds), do: {:ok, earth_age(seconds) / 1.8808158}
  def age_on(:jupiter, seconds), do: {:ok, earth_age(seconds) / 11.862615}
  def age_on(:saturn, seconds), do: {:ok, earth_age(seconds) / 29.447498}
  def age_on(:uranus, seconds), do: {:ok, earth_age(seconds) / 84.016846}
  def age_on(:neptune, seconds), do: {:ok, earth_age(seconds) / 164.79132}
  def age_on(_, _seconds), do: {:error, "not a planet"}

  defp earth_age(seconds), do: seconds / 31_557_600
end
